tensorflow/core/kernels/ctc_loss_op.cc

```diff
@@ -186,4 +220,152 @@ REGISTER_CPU(double);
+    OP_REQUIRES(ctx, ignore_longer_outputs_than_inputs_ == false,
+                errors::InvalidArgument("GPU CTCLossOp requires "
+                                        "ignore_longer_outputs_than_inputs_ to"
+                                        "be false"));
+
+    // Convert the labels_indices to labels_lengths
+    std::vector<int> labels_lengths(batch_size, 0);
+    DoHistogram<int64>(ctx, labels_indices, num_indices, batch_size,
+                       &labels_lengths);
+
+    StreamExecutor* executor = ctx->op_device_context()->stream()->parent();
+    se::dnn::DataType data_type = ToDataType<float>::value;
+
+    se::dnn::CtcLossDescriptor ctc_loss_desc;
+    std::unique_ptr<RnnStateTensorDescriptor> probs_desc;
+    std::unique_ptr<RnnStateTensorDescriptor> grads_desc;
+
+    auto probs_desc_s = executor->createRnnStateTensorDescriptor(
+        max_time, batch_size, num_classes, data_type);
+    OP_REQUIRES_OK(ctx, probs_desc_s.status());
+    probs_desc = probs_desc_s.ConsumeValueOrDie();
+
+    auto grads_desc_s = executor->createRnnStateTensorDescriptor(
+        max_time, batch_size, num_classes, data_type);
+    OP_REQUIRES_OK(ctx, grads_desc_s.status());
+    grads_desc = grads_desc_s.ConsumeValueOrDie();
+
+    absl::Span<const int32> labels_data;
+    absl::Span<const int32> labels_lengths_data;
+    absl::Span<const int32> input_lengths_data;
+    labels_data = absl::Span<const int32>(
+        labels_values->flat<int32>().data(), num_indices);
+    labels_lengths_data = absl::Span<const int32>(
+        labels_lengths.data(), batch_size);
+    input_lengths_data = absl::Span<const int32>(
+        seq_len->flat<int32>().data(), batch_size);
+
+    auto probs_data = StreamExecutorUtil::AsDeviceMemory<float>(*inputs);
+    auto costs_data = StreamExecutorUtil::AsDeviceMemory<float>(*loss);
+    auto grads_data = StreamExecutorUtil::AsDeviceMemory<float>(*gradient);
+
+    CudnnAllocatorInTemp workspace_allocator(ctx);
+
+    Stream* stream = ctx->op_device_context()->stream();
+    bool cudnn_launch_status =
+        stream
+            ->ThenCtcLoss(
+                *probs_desc, probs_data, labels_data, labels_lengths_data,
+                input_lengths_data, &costs_data, *grads_desc, &grads_data,
+                ctc_loss_desc, &workspace_allocator)
+            .ok();
+
+    if (!cudnn_launch_status) {
+      ctx->SetStatus(
+          errors::Internal("cuDNN CTCLoss launch failure"));
+    }
+  }
+
+ private:
+  bool preprocess_collapse_repeated_;
+  bool ctc_merge_repeated_;
+  bool ignore_longer_outputs_than_inputs_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(CTCLossOpGPU);
+};
+
+REGISTER_KERNEL_BUILDER(Name("CTCLossV2").Device(DEVICE_GPU)
+                                       .HostMemory("labels_indices")
+                                       .HostMemory("labels_values")
+                                       .HostMemory("sequence_length"),
+                        CTCLossOpGPU);
```

Comment:
Should we register the kernel even if cuDNN is older than 7.6.3 (and the kernel is guaranteed to fail)?